File: src/api/preview.js
'use strict';

const { Readable } = require('stream');

const STREAM_KEY_RE = /^[A-Za-z0-9_-]+$/;
const HLS_ASSET_SEGMENT_RE = /^[A-Za-z0-9._-]+$/;
const MAX_HLS_ASSET_PATH_CHARS = 512;
const MAX_HLS_ASSET_SEGMENTS = 16;
const HLS_PROXY_TIMEOUT_MS = 30000;
const MAX_HLS_MANIFEST_BYTES = 1024 * 1024;

function parseHlsAssetPath(rawAssetPath) {
    const assetPath = typeof rawAssetPath === 'string' && rawAssetPath.trim()
        ? rawAssetPath.trim()
        : 'index.m3u8';

    if (assetPath.length > MAX_HLS_ASSET_PATH_CHARS) return null;

    const segments = assetPath.split('/');
    if (
        segments.length === 0 ||
        segments.length > MAX_HLS_ASSET_SEGMENTS ||
        segments.some(
            (segment) =>
                !segment ||
                segment === '.' ||
                segment === '..' ||
                !HLS_ASSET_SEGMENT_RE.test(segment),
        )
    ) {
        return null;
    }

    return {
        encodedPath: segments.map((segment) => encodeURIComponent(segment)).join('/'),
        rawPath: assetPath,
    };
}

function buildForwardRequestHeaders(req) {
    const headers = {};
    const ifNoneMatch = req.headers['if-none-match'];
    const ifModifiedSince = req.headers['if-modified-since'];
    const range = req.headers.range;

    if (typeof ifNoneMatch === 'string' && ifNoneMatch.trim()) {
        headers['if-none-match'] = ifNoneMatch;
    }
    if (typeof ifModifiedSince === 'string' && ifModifiedSince.trim()) {
        headers['if-modified-since'] = ifModifiedSince;
    }
    if (typeof range === 'string' && range.trim()) {
        headers.range = range;
    }

    return headers;
}

function copyAllowedUpstreamHeaders(upstreamResponse, res) {
    const passthroughHeaders = [
        'content-type',
        'cache-control',
        'etag',
        'last-modified',
        'accept-ranges',
        'content-range',
        'content-length',
    ];

    passthroughHeaders.forEach((headerName) => {
        const headerValue = upstreamResponse.headers.get(headerName);
        if (headerValue) res.setHeader(headerName, headerValue);
    });

    res.setHeader('x-content-type-options', 'nosniff');
}

function isManifestResponse(pathName, contentType) {
    return pathName.toLowerCase().endsWith('.m3u8') || /application\/(vnd\.apple\.mpegurl|x-mpegurl)/i.test(contentType || '');
}

function toNodeReadable(body) {
    if (!body) return null;
    if (typeof body.pipe === 'function') return body;
    if (typeof body.getReader === 'function') return Readable.fromWeb(body);
    if (typeof body[Symbol.asyncIterator] === 'function') return Readable.from(body);
    return Readable.from(body);
}

function runCleanupOnce(cleanup) {
    let cleanedUp = false;
    return () => {
        if (cleanedUp) return;
        cleanedUp = true;
        cleanup?.();
    };
}

async function readManifestBufferWithLimit({ body, maxBytes, abortController }) {
    if (!body) return Buffer.alloc(0);

    if (typeof body.getReader === 'function') {
        const reader = body.getReader();
        const chunks = [];
        let totalBytes = 0;

        try {
            while (true) {
                const { done, value } = await reader.read();
                if (done) break;

                const chunk = Buffer.isBuffer(value) ? value : Buffer.from(value);
                totalBytes += chunk.length;
                if (totalBytes > maxBytes) {
                    abortController?.abort();
                    await reader.cancel('manifest limit exceeded').catch(() => {});
                    return null;
                }

                chunks.push(chunk);
            }

            return Buffer.concat(chunks, totalBytes);
        } finally {
            try {
                reader.releaseLock();
            } catch (_) {
                // Ignore release failures after cancel/close.
            }
        }
    }

    const nodeStream = toNodeReadable(body);
    const chunks = [];
    let totalBytes = 0;

    try {
        for await (const value of nodeStream) {
            const chunk = Buffer.isBuffer(value) ? value : Buffer.from(value);
            totalBytes += chunk.length;
            if (totalBytes > maxBytes) {
                abortController?.abort();
                if (typeof nodeStream.destroy === 'function' && !nodeStream.destroyed) {
                    nodeStream.destroy();
                }
                return null;
            }

            chunks.push(chunk);
        }

        return Buffer.concat(chunks, totalBytes);
    } finally {
        if (
            nodeStream &&
            typeof nodeStream.destroy === 'function' &&
            !nodeStream.readableEnded &&
            !nodeStream.destroyed
        ) {
            nodeStream.destroy();
        }
    }
}

async function fetchUpstreamAsset({
    req,
    res,
    fetch,
    log,
    streamKey,
    assetPath,
    query,
    getMediamtxHlsBaseUrl,
    buildMediamtxPath,
}) {
    const upstreamUrl = `${getMediamtxHlsBaseUrl()}/${buildMediamtxPath(streamKey)}/${assetPath}${query || ''}`;
    const abortController = new AbortController();
    const timeout = setTimeout(() => abortController.abort(), HLS_PROXY_TIMEOUT_MS);
    const abortOnClientClose = () => {
        if (!res.writableEnded) abortController.abort();
    };
    res.on('close', abortOnClientClose);
    const cleanup = () => {
        clearTimeout(timeout);
        res.off('close', abortOnClientClose);
    };

    try {
        const upstreamResponse = await fetch(upstreamUrl, {
            headers: buildForwardRequestHeaders(req),
            signal: abortController.signal,
        });

        return {
            abortController,
            cleanup,
            upstreamResponse,
        };
    } catch (err) {
        cleanup();
        log('warn', 'HLS preview proxy upstream request failed', {
            streamKey,
            assetPath,
            error: err?.message || String(err),
        });
        return null;
    }
}

async function streamUpstreamResponse({
    upstreamResponse,
    res,
    pathName,
    abortController,
    cleanup,
    log,
}) {
    const finalize = runCleanupOnce(cleanup);
    const contentType = upstreamResponse.headers.get('content-type') || '';
    if (isManifestResponse(pathName, contentType)) {
        try {
            const buffer = await readManifestBufferWithLimit({
                body: upstreamResponse.body,
                maxBytes: MAX_HLS_MANIFEST_BYTES,
                abortController,
            });
            if (!buffer) {
                res.removeHeader('content-type');
                res.removeHeader('content-length');
                return res.status(502).json({ error: 'Preview manifest exceeds safe proxy size limit' });
            }
            return res.send(buffer);
        } catch (err) {
            log?.('warn', 'HLS preview proxy manifest read failed', {
                pathName,
                error: err?.message || String(err),
            });
            if (res.headersSent || res.writableEnded) {
                res.destroy(err);
                return;
            }
            res.removeHeader('content-type');
            res.removeHeader('content-length');
            return res.status(502).json({ error: 'Failed to read preview manifest' });
        } finally {
            finalize();
        }
    }

    if (!upstreamResponse.body) {
        try {
            return res.end();
        } finally {
            finalize();
        }
    }

    const bodyStream = toNodeReadable(upstreamResponse.body);
    bodyStream.once('error', (err) => {
        finalize();
        if (res.headersSent || res.writableEnded) {
            res.destroy(err);
            return;
        }
        res.status(502).json({ error: 'Failed to stream preview asset' });
    });
    res.once('close', finalize);
    res.once('finish', finalize);
    bodyStream.pipe(res);
}

function registerPreviewProxyRoutes({ app, fetch, log, getMediamtxHlsBaseUrl, buildMediamtxPath }) {
    async function proxyHlsAsset(req, res, rawAssetPath) {
        const streamKey = String(req.params.streamKey || '').trim();
        if (!STREAM_KEY_RE.test(streamKey)) {
            return res.status(400).json({ error: 'Invalid stream key' });
        }

        let parsedAsset = parseHlsAssetPath(rawAssetPath);
        if (!parsedAsset) {
            return res.status(400).json({ error: 'Invalid HLS asset path' });
        }

        const query = req.originalUrl.includes('?')
            ? req.originalUrl.slice(req.originalUrl.indexOf('?'))
            : '';
        const upstreamResult = await fetchUpstreamAsset({
            req,
            res,
            fetch,
            log,
            streamKey,
            assetPath: parsedAsset.encodedPath,
            query,
            getMediamtxHlsBaseUrl,
            buildMediamtxPath,
        });

        if (!upstreamResult) {
            return res.status(502).json({ error: 'Failed to fetch preview asset' });
        }

        const { abortController, cleanup, upstreamResponse } = upstreamResult;

        res.status(upstreamResponse.status);
        copyAllowedUpstreamHeaders(upstreamResponse, res);
        return streamUpstreamResponse({
            abortController,
            cleanup,
            log,
            upstreamResponse,
            res,
            pathName: parsedAsset.rawPath,
        });
    }

    app.get('/preview/hls/:streamKey', async (req, res) => {
        await proxyHlsAsset(req, res, 'index.m3u8');
    });

    app.get('/preview/hls/:streamKey/*assetPath', async (req, res) => {
        const wildcard = req.params.assetPath;
        const assetPath = Array.isArray(wildcard)
            ? wildcard.join('/')
            : wildcard || 'index.m3u8';
        await proxyHlsAsset(req, res, assetPath);
    });
}

module.exports = {
    registerPreviewProxyRoutes,
};

File: internal/api/preview.go
package api

import (
	"io"
	"net/http"
	"net/url"
	"regexp"
	"strings"
	"time"

	"github.com/go-chi/chi/v5"

	"restream/internal/apputils"
	"restream/internal/mediamtx"
)

var (
	hlsAssetSegmentRE    = regexp.MustCompile(`^[A-Za-z0-9._-]+$`)
	maxHLSAssetPathChars = 512
	maxHLSAssetSegments  = 16
	hlsProxyTimeout      = 30 * time.Second
	maxHLSManifestBytes  = int64(1024 * 1024)

	forwardedResponseHeaders = []string{
		"content-type",
		"cache-control",
		"etag",
		"last-modified",
		"accept-ranges",
		"content-range",
		"content-length",
	}
)

type parsedHLSAsset struct {
	encodedPath string
	rawPath     string
}

func parseHLSAssetPath(rawAssetPath string) *parsedHLSAsset {
	assetPath := strings.TrimSpace(rawAssetPath)
	if assetPath == "" {
		assetPath = "index.m3u8"
	}
	if len(assetPath) > maxHLSAssetPathChars {
		return nil
	}
	segments := strings.Split(assetPath, "/")
	if len(segments) == 0 || len(segments) > maxHLSAssetSegments {
		return nil
	}
	encodedParts := make([]string, 0, len(segments))
	for _, seg := range segments {
		if seg == "" || seg == "." || seg == ".." || !hlsAssetSegmentRE.MatchString(seg) {
			return nil
		}
		encodedParts = append(encodedParts, url.PathEscape(seg))
	}
	return &parsedHLSAsset{
		encodedPath: strings.Join(encodedParts, "/"),
		rawPath:     assetPath,
	}
}

func isManifestResponse(pathName, contentType string) bool {
	return strings.HasSuffix(strings.ToLower(pathName), ".m3u8") ||
		strings.Contains(strings.ToLower(contentType), "application/vnd.apple.mpegurl") ||
		strings.Contains(strings.ToLower(contentType), "application/x-mpegurl")
}

func copyAllowedUpstreamHeaders(upstreamResp *http.Response, w http.ResponseWriter) {
	for _, h := range forwardedResponseHeaders {
		if v := upstreamResp.Header.Get(h); v != "" {
			w.Header().Set(h, v)
		}
	}
	w.Header().Set("x-content-type-options", "nosniff")
}

func clearForwardedHeaders(w http.ResponseWriter) {
	for _, h := range forwardedResponseHeaders {
		w.Header().Del(h)
	}
}

func buildForwardRequestHeaders(r *http.Request) http.Header {
	headers := make(http.Header)
	if v := r.Header.Get("if-none-match"); strings.TrimSpace(v) != "" {
		headers.Set("if-none-match", v)
	}
	if v := r.Header.Get("if-modified-since"); strings.TrimSpace(v) != "" {
		headers.Set("if-modified-since", v)
	}
	if v := r.Header.Get("range"); strings.TrimSpace(v) != "" {
		headers.Set("range", v)
	}
	return headers
}

var hlsProxyClient = &http.Client{
	Timeout: hlsProxyTimeout,
}

// RegisterPreviewAPI mounts the /preview/hls/:streamKey proxy.
func RegisterPreviewAPI(r chi.Router) {
	r.Get("/preview/hls/{streamKey}", func(w http.ResponseWriter, r *http.Request) {
		proxyHLSAsset(w, r, "index.m3u8")
	})

	r.Get("/preview/hls/{streamKey}/*", func(w http.ResponseWriter, r *http.Request) {
		wildcard := chi.URLParam(r, "*")
		if wildcard == "" {
			wildcard = "index.m3u8"
		}
		proxyHLSAsset(w, r, wildcard)
	})
}

func proxyHLSAsset(w http.ResponseWriter, r *http.Request, rawAssetPath string) {
	streamKey := strings.TrimSpace(chi.URLParam(r, "streamKey"))
	if apputils.ValidateStreamKey(streamKey, "Stream key") != "" {
		jsonError(w, http.StatusBadRequest, "Invalid stream key")
		return
	}

	parsed := parseHLSAssetPath(rawAssetPath)
	if parsed == nil {
		jsonError(w, http.StatusBadRequest, "Invalid HLS asset path")
		return
	}

	query := ""
	if idx := strings.Index(r.RequestURI, "?"); idx >= 0 {
		query = r.RequestURI[idx:]
	}

	hlsBase := mediamtx.GetHLSBaseURL()
	mtxPath := mediamtx.BuildPath(streamKey)
	upstreamURL := hlsBase + "/" + mtxPath + "/" + parsed.encodedPath + query

	req, err := http.NewRequestWithContext(r.Context(), http.MethodGet, upstreamURL, nil)
	if err != nil {
		jsonError(w, http.StatusInternalServerError, "Failed to build upstream request")
		return
	}
	for k, vals := range buildForwardRequestHeaders(r) {
		for _, v := range vals {
			req.Header.Set(k, v)
		}
	}

	resp, err := hlsProxyClient.Do(req)
	if err != nil {
		apputils.Log("warn", "HLS preview proxy upstream request failed", map[string]interface{}{
			"streamKey": apputils.MaskToken(streamKey),
			"assetPath": parsed.rawPath,
			"error":     err.Error(),
		})
		jsonError(w, http.StatusBadGateway, "Failed to fetch preview asset")
		return
	}
	defer resp.Body.Close()

	copyAllowedUpstreamHeaders(resp, w)
	w.WriteHeader(resp.StatusCode)

	if isManifestResponse(parsed.rawPath, resp.Header.Get("content-type")) {
		limited := io.LimitReader(resp.Body, maxHLSManifestBytes+1)
		buf, err := io.ReadAll(limited)
		if err != nil {
			if w.Header().Get("content-type") != "" {
				return
			}
			clearForwardedHeaders(w)
			return
		}
		if int64(len(buf)) > maxHLSManifestBytes {
			clearForwardedHeaders(w)
			http.Error(w, `{"error":"Preview manifest exceeds safe proxy size limit"}`,
				http.StatusBadGateway)
			return
		}
		_, _ = w.Write(buf)
		return
	}

	_, _ = io.Copy(w, resp.Body)
}
